Quantized addition refactor

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24217 AddReLU module
* #23971 Quantized addition with 'out'
* **#23956 Quantized addition refactor**

Differential Revision: [D16692445](https://our.internmc.facebook.com/intern/diff/D16692445)

Commit:
[1/2] Add out variant

Commit:
Update on "[1/2] Add out variant"


Differential Revision: [D16692445](https://our.internmc.facebook.com/intern/diff/D16692445)

Commit:
Update on "[1/2] Add out variant"


Differential Revision: [D16692445](https://our.internmc.facebook.com/intern/diff/D16692445)

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -9,48 +9,62 @@
 namespace native {
 namespace {
+
+inline Tensor requantize_into_int32(Tensor qa) {
```
not used?

jamesr66a replied:
Also did you actually mean to take the tensor by value? Probably better to take a const ref so we don't hit atomic refcount bumps

zafartahirov replied:
@jamesr66a I was told before that with the newest changes to at::Tensor, there is no need to pass it by const ref. I might have misunderstood something.

jamesr66a replied:
Tensor is still a thin wrapper around intrusive_ptr<TensorImpl> which has--not one--but TWO atomic refcounts that need to be updated on copies. This is a high cost because of synchronization overheads and should be avoided if you're only reading properties from a tensor.

See my recent note about profiling with flame graphs. We've spent a lot of time recently removing unnecessary Tensor copies to reduce this overhead

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -9,48 +9,62 @@
+ public:
+  Tensor operator()(at::Tensor qa, at::Tensor qb,
+                    double scale, int64_t zero_point) {
+    TORCH_CHECK(qa.numel() == qb.numel(),
```
btw, why we don't support broadcasting? if it's using TensorIterator it should just work :)

zafartahirov replied:
It should. The problem is that we never discussed/defined the behavior of the broadcasting in the context of the quantization. I will send a separate PR with broadcasting implementation, and we can discuss it there.

raghuramank100 replied:
For quantized add, we should support broadcasting as long as both the operands are quantized and satisfy the conditions (i.e same datatype, qscheme is not per_channel_quant)

zafartahirov replied:
Will do as a separate diff

zafartahirov replied:
Created an issue to keep track of this feature: #24210

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -9,48 +9,62 @@
+  auto iter = TensorIterator::binary_op(out, self, other);
+  AT_DISPATCH_QINT_TYPES(out.scalar_type(), "qadd", [&]() {
+    cpu_kernel(iter, [&](scalar_t a, scalar_t b) -> scalar_t {
+      const auto da = at::dequantize_val(self.q_scale(), self.q_zero_point(), a);
```
calls to q_scale / q_zero_point go through dispatch - call them once in the beginning

jamesr66a commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -9,48 +9,62 @@
+    return qa;
+  }
+  const auto a = qa.dequantize();
+  return at::quantize_linear(a, /*scale=*/1.0, /*zero_point=*/0, kQInt32);;
```
Double semicolon

jamesr66a commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -9,48 +9,62 @@
-    auto c = at::empty_like(a);
-    auto iter = TensorIterator::binary_op(c, a, b);
+Tensor _add_out(Tensor& out, const Tensor& self, const Tensor& other) {
+  long zero_point = out.q_zero_point();
```
s/long/int64_t/

long is platform-dependent and we use int64_t for everything in pytorch

jamesr66a commented on c10/core/ScalarType.h:
```diff
@@ -269,7 +269,7 @@ static inline ScalarType promoteTypes(ScalarType a, ScalarType b) {
   static constexpr ScalarType _promoteTypesLookup[static_cast<int>(
       ScalarType::NumOptions)][static_cast<int>(ScalarType::NumOptions)] = {
-      /* u1  i1  i2  i4  i8  f2  f4  f8  c2  c4  c8  b1 */
+      /*        u1  i1  i2  i4  i8  f2  f4  f8  c2  c4  c8  b1 */
```
Is this from autoformatting?

zafartahirov replied:
Accidental commit -- I was double checking the promotions -- forgot to cleanup.

jamesr66a commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -9,48 +9,62 @@
-} // namespace
-} // namespace native
-} // namespace at
+static auto registry = c10::RegisterOperators()
```
Did you change anything here?

zafartahirov replied:
Yes, minor refactor in the formatting to allow better usage of limited line-length

ZolotukhinM commented:
Please note that when this got committed you won't see the stack in the commit message. You'll only see "[1/2] Add out variant" which is not the most informative commit message in the world :)

Commit:
Update on "Quantized addition refactor"


Differential Revision: [D16692445](https://our.internmc.facebook.com/intern/diff/D16692445)

Commit:
Update on "Quantized addition refactor"


Differential Revision: [D16692445](https://our.internmc.facebook.com/intern/diff/D16692445)

Commit:
Update on "Quantized addition refactor"


Differential Revision: [D16692445](https://our.internmc.facebook.com/intern/diff/D16692445)

Commit:
Update on "Quantized addition refactor"


Differential Revision: [D16692445](https://our.internmc.facebook.com/intern/diff/D16692445)

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -9,48 +9,66 @@
+  TORCH_CHECK(qa.qscheme() == kPerTensorAffine ||
+              qa.qscheme() == kPerTensorSymmetric,
+              "Only per tensor quantization is suuported in Add.");
+  TORCH_CHECK(qa.qscheme() == qb.qscheme(),
```
We can mix and match qschemes for a and b, so we should only check:
```
TORCH_CHECK(qa.qscheme() == kPerTensorAffine ||
              qa.qscheme() == kPerTensorSymmetric,
              "Only per tensor quantization is supported in Add.");
TORCH_CHECK(qb.qscheme() == kPerTensorAffine ||
              qb.qscheme() == kPerTensorSymmetric,
              "Only per tensor quantization is supported in Add.");
```

zafartahirov replied:
Will send it as a separate diff

zafartahirov replied:
Created an issue to keep track of this task: #24215
